scrapy/core/downloader/handlers/negotiation.py

```diff
@@ -0,0 +1,284 @@
+from collections import deque
+from urllib.parse import urldefrag
+
+from twisted.internet.base import ReactorBase
+from twisted.internet.defer import Deferred
+from twisted.internet.interfaces import IHandshakeListener
+from twisted.internet.protocol import Factory, Protocol
+from twisted.internet.protocol import connectionDone
+from twisted.web._newclient import HTTP11ClientProtocol
+from twisted.web.client import BrowserLikePolicyForHTTPS, _StandardEndpointFactory, URI
+from typing import Deque, Dict, List, Optional, Tuple, Union
+from zope.interface import implementer
+
+from scrapy.core.downloader.contextfactory import AcceptableProtocolsContextFactory, load_context_factory_from_settings
+from scrapy.core.downloader.handlers.http11 import HTTP11DownloadHandler
+from scrapy.core.downloader.handlers.http2 import H2DownloadHandler
+from scrapy.core.http2.protocol import H2ClientProtocol
+from scrapy.http.request import Request
+from scrapy.settings import Settings
+from scrapy.spiders import Spider
+from scrapy.utils.datatypes import LocalCache
+from scrapy.utils.python import to_bytes
+
+SUPPORTED_PROTOCOLS = ('h2', 'http/1.1')
+
+
+class UnsupportedNegotiatedProtocol(Exception):
```

Comment:
If we trust the [coverage data](https://codecov.io/gh/scrapy/scrapy/pull/4770/diff), none of the code below that creates an instance of this class gets executed during tests.

Reply:
Yes. I am not sure although how to test this, I added this in case `Twisted` adds a support for HTTP/3 Server or a request is made to the website which chooses HTTP/2 during ALPN or NPN. However, this is a very rare case we are sending `SUPPORTED_PROTOCOLS` during negotiation such that only one of HTTP/1 or HTTP/2 is chosen.

Reply:
For the HTTP/3 case, I think the ideal solution would be to add a mock server which only supports HTTP/3, and test that scenario.

We could alternatively [mock](https://docs.python.org/3/library/unittest.mock.html) our code in the tests, to behave as it would in the scenarios we devise.

> a request is made to the website which chooses HTTP/2 during ALPN or NPN

Could you elaborate? How does this scenario lead to `UnsupportedNegotiatedProtocol`, given `SUPPORTED_PROTOCOLS = ('h2', 'http/1.1')`?

Reply:
> Could you elaborate? How does this scenario lead to `UnsupportedNegotiatedProtocol`, given SUPPORTED_PROTOCOLS = ('h2', 'http/1.1')?

Thinking about it. I realize that as long as we specify the `SUPPORTED_PROTOCOLS = ('h2', 'http/1.1')` the exception will never be raised. As a result either no protocol will be negotiated or one of the two. However, I'm not sure about the behaviour of the download handler when no protocol is negotiated. I will add a test for that case now. 

Should we remove the `UnsupportedNegotiatedProtocol`? I do feel now that it will never be raised as we always specify a list of protocols to be negotiated.

Reply:
> Should we remove the `UnsupportedNegotiatedProtocol`? I do feel now that it will never be raised as we always specify a list of protocols to be negotiated.

Yes, I think that makes sense. Thanks!